Source/WebCore/platform/graphics/ImageBuffer.cpp

```diff
@@ -126,9 +127,39 @@ RefPtr<NativeImage> ImageBuffer::sinkIntoNativeImage(RefPtr<ImageBuffer> imageBu
     return imageBuffer->sinkIntoNativeImage();
 }
 
-RefPtr<Image> ImageBuffer::sinkIntoImage(RefPtr<ImageBuffer> imageBuffer, PreserveResolution preserveResolution)
+RefPtr<Image> ImageBuffer::copyImage(BackingStoreCopy copyBehavior, PreserveResolution preserveResolution) const
 {
-    return imageBuffer->sinkIntoImage(preserveResolution);
+    RefPtr<NativeImage> image;
+    if (resolutionScale() == 1 || preserveResolution == PreserveResolution::Yes)
+        image = copyNativeImage(copyBehavior);
+    else {
+        auto copyBuffer = context().createImageBuffer(logicalSize(), 1.f, colorSpace());
+        if (!copyBuffer)
+            return nullptr;
+        copyBuffer->context().drawImageBuffer(*const_cast<ImageBuffer*>(this), FloatPoint { }, CompositeOperator::Copy);
+        image = ImageBuffer::sinkIntoNativeImage(WTFMove(copyBuffer));
+    }
+    if (!image)
+        return nullptr;
+    return BitmapImage::create(WTFMove(image));
+}
+
+RefPtr<Image> ImageBuffer::sinkIntoImage(RefPtr<ImageBuffer> source, PreserveResolution preserveResolution)
+{
+    RefPtr<NativeImage> image;
+    if (source->resolutionScale() == 1 || preserveResolution == PreserveResolution::Yes)
+        image = sinkIntoNativeImage(WTFMove(source));
+    else {
+        auto copySize = source->logicalSize();
+        auto copyBuffer = source->context().createImageBuffer(expandedIntSize(copySize), 1.f, source->colorSpace());
+        if (!copyBuffer)
+            return nullptr;
+        drawConsuming(WTFMove(source), copyBuffer->context(), FloatRect { { }, copySize }, FloatRect { 0, 0, -1, -1 }, CompositeOperator::Copy);
```

Comment:
Because we want to replace the destination pixels with source pixels, without blending. We know the destination is empty, so we don't want to read the destination pixels redundantly.